### PR TITLE
Autogenerate Scheme bindings against `stdlib.h`

### PR DESCRIPTION
**Pull request by mkeeter**
This is part 2 of the plan for ~world domination~ a canonical, easy-to-bind standard library (part 1 was #383).

Previously, all of the shapes in the standard library were defined in various Scheme files (`shapes.scm`, etc).

As of this PR, they're all defined as C++ functions in [`stdlib_impl.cpp`](https://github.com/libfive/libfive/blob/autogen-scheme-bindings/libfive/stdlib/stdlib_impl.cpp).

The standard library header at [`stdlib.h`](https://github.com/libfive/libfive/blob/autogen-scheme-bindings/libfive/stdlib/stdlib.h) is parsed and used to **generate**
- `stdlib_impl.hpp` (C++ header)
- `stdlib.cpp` (thin layer which converts from C types to C++ types, calling into `stdlib_impl.cpp`)
- `shapes/transforms/csg/text.scm`, which are now purely wrappers around FFI calls

This is neat, because now that we have infrastructure for generating _Scheme_ bindings, we can easily reuse it to generate _other_ bindings ([hissssssss](http://bash.org/?400459)).

I still need to go through and make sure that I ported all shapes correctly, because this was a lot of manual Scheme → C++ translation, and other testing would be welcome.

### CODE REVIEW

**Comment by koalahedron** [comment]
Compiles and runs OK on Ubuntu 18.04.2.  I tried all the models in ```./studio/examples```.  Only the koala model didn't work:

```
wrong-number-of-args: Wrong number of arguments to #<procedure scale-xyz (t s center)>

Stack trace:
0: (eval (koalahedron (truncated-icosohedron) ($bracket-list$ 0 0 0) +truncated-icosohedron-circumradius+) #<module (#{ g77}#) 55a0962c1140>)
1: 125000000
```

I'm going to try some of my more complicated models next.

Have you thought about doing a comprehensive unit test suite for the C-API?  For example making sure rotations, etc. work as they are described.  Another testing technique that has paid off for me has been adding a unit test for each verified bug as a regression test.  This could be done retroactively.  Test coverage reporting using the compiler is also a good way to gauge how much as been tested.

**Comment by mkeeter** [comment]
Thanks, Mr. Koala!  I've pushed a fix and added a missing definition of `pi`, as well as a few unit tests that would have caught this.

(There are already C API tests in `test/api.cpp` and Guile bindings tests in `test/guile.cpp`, but the Guile tests aren't totally comprehensive)

**Comment by mkeeter** [comment]
Looks like `attract` / `repel` are broken, so converting to draft.

**Comment by mkeeter** [comment]
Okay, looks good now!

**Comment by koalahedron** [comment]
Thanks for the updates.

I was thinking of doing automated language bindings to a C API with a generated approach using a spec.  An XML or JSON file would have definitions, and the binding author could traverse the file--almost every language has a parser for one or both.  I've had issues with C header file groveling (as the Common Lisp people call it).  You would also have a nice formal spec for interop.

For example OpenGL has an [XML spec](https://github.com/KhronosGroup/OpenGL-Registry/blob/6bbebc91b51c75687b907bacc867f1e1e3a92da9/xml/gl.xml).

A number of bindings have been made to OpenGL using this technique (I think Guile, for example).  I'm sure there are tradeoffs either way--but just FYI.

G'day mate.

File: libfive/stdlib/stdlib_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,1379 @@
...
+                             const auto y = Tree::Y(); (void)y; \
+                             const auto z = Tree::Z(); (void)z; ;
+////////////////////////////////////////////////////////////////////////////////
+// Operator overloads for C vecs
+TreeVec2 operator+(const TreeVec2& a, const TreeVec2& b) {
+    return TreeVec2{a.x + b.x, a.y + b.y};
+}
+TreeVec2 operator-(const TreeVec2& a, const TreeVec2& b) {
+    return TreeVec2{a.x - b.x, a.y - b.y};
+}
```
Comment (SGSSGene) [inline libfive/stdlib/stdlib_impl.cpp:24]: Maybe for completeness also add a negation operator for TreeVec2
```suggestion
}
TreeVec2 operator-(const TreeVec2& a) {
    return TreeVec2{-a.x, -a.y};
}
```

**Comment by SGSSGene** [comment]
First look at it, it looks nice. I will try it out later tonight.

**Comment by mkeeter** [comment]
I've gone through and done a cursory test of each space, so this is good to go – if it breaks things, I'm sure someone will let me know 😁

**Pull request merged**